File: app/src/main/java/com/hxh/kt10/java/LazyThreadSafeDoubleCheck.java
package com.hxh.kt10.java;

/**
 * Created by HXH on 2017/6/1 0001.
 * 懒加载单例(线程安全的-双重空检查)
 */

public class LazyThreadSafeDoubleCheck {
    private static volatile LazyThreadSafeDoubleCheck INSTANCE;

    private LazyThreadSafeDoubleCheck() {

    }

    public static LazyThreadSafeDoubleCheck getInstance() {
        if (INSTANCE == null) {
            synchronized (LazyThreadSafeDoubleCheck.class) {
                if (INSTANCE == null) {
                    // 初始化时分为实例化和赋值两步, 尽管我们把这一步写成下面的语句,
                    // 但Java虚拟机并不保证其他线程『眼中』这两步的顺序究竟是怎么样的
                    INSTANCE = new LazyThreadSafeDoubleCheck();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/hxh/kt10/kotlin/LazyThreadSafeDoubleCheck.kt
package com.hxh.kt10.kotlin

/**
 * Created by HXH on 2017/6/1 0001.
 *
 */
class LazyThreadSafeDoubleCheck private constructor() {

    companion object {
        val instance by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            LazyThreadSafeDoubleCheck()
        }

        private @Volatile var instance2: LazyThreadSafeDoubleCheck? = null

        fun get():LazyThreadSafeDoubleCheck{
            if (instance2 == null) {
                synchronized(this) {
                    if (instance2 == null) {
                        instance2 = LazyThreadSafeDoubleCheck()
                    }
                }
            }
            return instance2!!
        }
    }
}

fun main(args: Array<String>) {
    LazyThreadSafeDoubleCheck.instance.sayHello()
    LazyThreadSafeDoubleCheck.get().sayHello()
}

fun LazyThreadSafeDoubleCheck.sayHello(){
    println("hello")
}